primary_beam_correction.py

```diff
@@ -0,0 +1,296 @@
+    Making beam pattern using the DEEP2 paper
+    Input : Radial seperation array and path to the fits file
+    Output : Beam pattern
+    '''
+    if c_freq:
+        rho = separation_rad
+        c = c_freq*1e6
+        # Degrees to radians conversion
+        conv = np.pi / 180.
+        # converting arcminutes to radians
+        v_beam_rad = 89.5 / 60. * conv
+        h_beam_rad = 86.2 / 60. * conv
+        # Taking the Geometric mean for the veritical and horizontal cut through the beam.
+        vh_beam_mean = np.sqrt(v_beam_rad*h_beam_rad)
+        flux_density = []
+        nu = c/1.e9
+        theta_b = vh_beam_mean*(1./nu)
+        ratio = rho/theta_b
+        num = np.cos(1.189*np.pi*ratio)
+        dem = 1-4*(1.189*ratio)**2
+        a_b = (num/dem)**2
+        flux_density.append(a_b)
+    else:
+        rho = separation_rad
+        c = central_freq(path)
+        # Degrees to radians conversion
+        conv = np.pi / 180.
+        # converting arcminutes to radians
+        v_beam_rad = 89.5 / 60. * conv
+        h_beam_rad = 86.2 / 60. * conv
+        # Taking the Geometric mean for the veritical and horizontal cut through the beam.
+        vh_beam_mean = np.sqrt(v_beam_rad*h_beam_rad)
+        flux_density = []
+        for nu in c:
+            nu = nu/1.e9
+            theta_b = vh_beam_mean*(1./nu)
+            ratio = rho/theta_b
+            num = np.cos(1.189*np.pi*ratio)
+            dem = 1-4*(1.189*ratio)**2
+            a_b = (num/dem)**2
+            flux_density.append(a_b)
+    return flux_density
+
+
+def mad(data):
+    '''
+    Calculating meadian absolute deviation (MAD).
+    '''
+    MAD_TO_SD = 1.4826
+    med = np.nanmedian(data)
+    dev = np.abs(data - med)
+    return med, MAD_TO_SD * np.nanmedian(dev)
+
+
+def standard_deviation(data):
+    '''
+    Calcalating the standard deviation of each frequency plane using MAD.
+    Returns: Weights per each frequency plane
+    '''
+    diff = 1
+    i = 0
+    data = data[data != 0.0]
+    if len(data) == 0:
+        return float(0.0), float(0.0)
+    med, sd = mad(data)
+    while diff > 0:
+        i += 1
+        if i > 50:
+            return sd
+        old_sd = sd
+        old_len = len(data)
+        cut = np.abs(data - med) < 5.0 * sd
+        if np.all(~cut):
+            return sd
+        data = data[cut]
+        sd = mad(data)[1]
+        diff = old_len - len(data)
+        if sd == 0.0:
+            return old_sd
+    return 1/(sd)**2
+
+
+def weigted_average(arr, weights):
+    '''
+    Computing weighted average of all the frequency planes.
+    '''
+    wt_average = np.average(arr, weights=weights, axis=0)
+    return wt_average
+
+
+def primary_beam_correction(beam_pattern, path):
+    '''
+    Correcting for the primary beam effects.
+    '''
+    raw_image = read_fits(path)
+    raw_img = []
+    try:
+        nterm = raw_image.header['NTERM']
+    except KeyError:
+        nterm = _get_value_from_history('NTERM', raw_image.header)
```

Comment:
Why this? Does NTERM ever appear in the HISTORY of a FITS header?

Reply:
I was confusing NTERM with NSPEC but I have corrected that.